cmd/snap-update-ns/secure_bindmount.go

```diff
@@ -0,0 +1,84 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2018 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package main
+
+import (
+	"os"
+	"syscall"
+)
+
+// SecureBindMount performs a bind mount between two absolute paths
+// containing no symlinks, using a private stash directory as an
+// intermediate step
+func SecureBindMount(sourceDir, targetDir string, flags uint, stashDir string) error {
+	// Save source directory, since we use chdir internally
+	cwd, err := os.Getwd()
+	if err != nil {
+		return err
+	}
+	defer os.Chdir(cwd)
+
+	// Step 1: acquire file descriptors representing the source
+	// and destination directories, ensuring no symlinks are
+	// followed.
+	sourceFd, err := secureOpenPath(sourceDir)
+	if err != nil {
+		return err
+	}
+	defer sysClose(sourceFd)
+	targetFd, err := secureOpenPath(targetDir)
+	if err != nil {
+		return err
+	}
+	defer sysClose(targetFd)
+
+	// Step 2: chdir to the source, and bind mount "." to the stash dir
+	bindFlags := syscall.MS_BIND | (flags & syscall.MS_REC)
+	if err := sysFchdir(sourceFd); err != nil {
+		return err
+	}
+	if err := sysMount(".", stashDir, "", uintptr(bindFlags), ""); err != nil {
+		return err
+	}
+	defer sysUnmount(stashDir, syscall.MNT_DETACH)
+
+	// Step 3: optionally change to readonly
+	if flags&syscall.MS_RDONLY != 0 {
+		remountFlags := syscall.MS_REMOUNT | syscall.MS_BIND | syscall.MS_RDONLY
+		if flags&syscall.MS_REC != 0 {
+			remountFlags |= syscall.MS_REC
+		}
+		if err := sysMount("none", stashDir, "", uintptr(remountFlags), ""); err != nil {
+			return err
+		}
+	}
+
+	// Step 4: chdir to the destination, and move mount the stash to "."
+	if err := sysFchdir(targetFd); err != nil {
+		return err
+	}
+	// Ideally this would be a move rather than a second bind, but
+	// that fails for shared mount namespaces
+	if err := sysMount(stashDir, ".", "", uintptr(bindFlags), ""); err != nil {
+		return err
+	}
```

Comment:
I was using MNT_DETACH to minimise chance of blocking on the unmount.  When I tested things out, there is no reference to the stash mount point in `/proc/self/mountinfo` after the process completes, so I don't have any reason to believe the stash mount sticks around: the final mounts at the target reference the original source location directly.

We don't try to delete the stash directory because it is provided by the caller: it could easily delete it after performing all the bind mounts.

Reply:
The stash mount would stick around in the namespace of the process but yeah, we should detach it. Not just because there's a chance of blocking but also because if we ever do recursive bind mounts, there's no other way than MNT_DETACH in practice.